tests: lib: peer_manager: add tests for security dispatcher

eivindj-nordic opened the pull request:
Add tests for security dispatcher.
One fix regarding conn state return.

Some tests are disabled (#if 0) as they are n0ot working out of the box and likely needing central.

eivindj-nordic committed:
lib: add conn_state

Add Connection state library.

Signed-off-by: Eivind Jølsgard <eivind.jolsgard@nordicsemi.no>
Co-authored-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

MirkoCovizzi committed:
lib: add Peer Manager

Adds the Peer Manager.

List of changes compared to nrf5 SDK:

* Restructured and moved headers to dedicated folders

* Applied `clang-format-15` to each file

* Updated all licenses

* Replaced `ret_code_t` with `uint32_t`

* Connection state flags are not present in `ble_conn_state` so
  use the Kconfig `BLE_CONN_STATE_USER_FLAG_COUNT` instead.
  Convert the type `ble_conn_state_user_flag_id_t` to `int`.

* The configurable values defined in `peer_manager.packsc` have
  been converted to Kconfig options with a python script, and
  replaced in the code.

* Reformatted all Doxygen to fit NCS style

* Removed nrf5 SDK logging in favor of Zephyr logging

* DOS to unix line endings

* Fix Checkpatch

* Use Zephyr assert in `sdk_macros.h`

* All includes of headers that do not exist in Bare Metal have
  been removed.

* Changed `NRF_ERROR_RESOURCES` to `NRF_ERROR_DATA_SIZE`.
  This is in order to use `NRF_ERROR_RESOURCES` in stead of
  `NRF_ERROR_STORAGE_FULL`.

* Changed `NRF_ERROR_STORAGE_FULL` to `NRF_ERROR_RESOURCES`.

* Put central logic compilation behind Kconfig flag

* Add some extra build dependencies

* Update GATT Cache Manager by porting `nrf_mtx` logic

* Update `peer_id` and `pm_buffer` to use `atomic_t`

Signed-off-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

MirkoCovizzi committed:
auth_status_tracker: use bm_timer

Fixup.

Signed-off-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

MirkoCovizzi committed:
Patch workflows for integration development

Path workflows for integration development.

Signed-off-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

MirkoCovizzi committed:
peer_manager_handler: use bm_timer

Fixup.

Signed-off-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

anhmolt committed:
lib: peer_manager: port lesc module to use PSA APIs

Update the LE secure part of the peer_manager to use PSA APIs.
In SDK5 they were using the nrf_crypto libraries.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

eivindj-nordic committed:
tests: lib: peer_manager: add tests for security dispatcher

Add tests for security dispatcher.
One fix regarding conn state return.

Signed-off-by: Eivind Jølsgard <eivind.jolsgard@nordicsemi.no>

MirkoCovizzi commented on tests/lib/peer_manager/security_dispatcher/CMakeLists.txt:
```diff
@@ -0,0 +1,51 @@
+  CONFIG_BLE_CONN_STATE_USER_FLAG_COUNT=24
+)
+
+
```
Extra newline

eivindj-nordic replied:
I have excluded the test from checkpatch as it is not planned to be merged.

eivindj-nordic replied:
Update: file is no longer excluded from checkpatch as we plan to merge as is after all.

MirkoCovizzi reviewed (commented):
How much was it changed compared to the original?

oyvindronningstad commented on tests/lib/peer_manager/security_dispatcher/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+
+project(unit_test_ble_qwr)
```
nit: Copy paste error?

eivindj-nordic replied:
Indeed.

anhmolt committed:
lib: peer_manager: id_manager: do not use conn_handle as array index

Connection handle values cannot be used to lookup arrays.
Fix this by using the nrf_sdh_ble_idx_get function to get
an index that can be used in place of the conn_handle
when indexing arrays.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

eivindj-nordic commented:
> How much was it changed compared to the original?

Mainly updates to CMakeLists, aligning to new cmock style, using extern variables instead of including the c-file under test itself, have more tests depending on central behind the if.  Same type of things as in #6.

MirkoCovizzi committed:
peer_manager: Use proper logging macro

Fixup.

Signed-off-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

MirkoCovizzi committed:
peer_manager: nordic_common: import BYTES_TO_WORDS macro

Fixup.

Signed-off-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

MirkoCovizzi committed:
peer_manager: add missing includes

Fixup.

Signed-off-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

MirkoCovizzi committed:
ble_hrs: add missing Kconfig

Fixup.

Signed-off-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

MirkoCovizzi committed:
ble_hrs: use correct function

Fixup.

Signed-off-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

MirkoCovizzi committed:
peer_manager: disable APP_ERROR_CHECK

Fixup.

Signed-off-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

anhmolt committed:
ble_hrs: add crypto configuration for LE secure connection support

Add the required crypto configuration for using the LE Secure
Connection module of the peer manager. This is done to
show what is required.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

anhmolt committed:
lib: peer_manager: changes to get rid of nordic_common.h

Change peer_manager so it no longer depends on macros
from nordic_common.h. Some are substituted with macros
from the zephyr utils module. Other dependencies are
solved by adding the macro to the source files in question.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

anhmolt committed:
lib: peer_manager: fix wrong macro description for an unused macro

 The description of the PM_REMOTE_DB_N_SERVICES macro is wrong.
 Doxygen is updated.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

anhmolt committed:
lib: peer_manager: Fix a logging line with too many parameters

Three format specifiers for outputting values existed, but
only two arguments were given. Remove the last format specifier,
which also seems like a duplicate of the first one.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

anhmolt committed:
lib: peer_manager: fix the APP_ERROR_CHECK macro throwing warnings

Give the APP_ERROR_CHECK macro a parameter so it will stop
to generate warnings.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

anhmolt committed:
lib: peer_manager: fix missing ALIGN_NUM macro

The ALIGN_NUM macro was defined in sdk5 file app_util.h.
The plan is to not port it. Use zephyr util macros
inplace of the ALIGN_NUM macro.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

MirkoCovizzi committed:
peer_manager: move peer_manager_handler.h

Fixup.

Signed-off-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

MirkoCovizzi committed:
peer_manager: update peer_data_storage

Fixup.

Signed-off-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

anhmolt committed:
lib: peer_manager: nrf_ble_lesc: remove sdk_macros header include

Rewrites the last part of the code using macro VERIFY_SUCCESS()
from sdk_macros.h and removes include sdk_macros.h.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

anhmolt committed:
lib: peer_manager: nrf_ble_lesc: fix array idx used as conn_handle

Fix one occurance where the array index from nrf_sdh_ble_idx_get()
was incorrectly used in place of connection handle.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>
